### PR TITLE
Implement token repair service #5135

### LINKED ISSUE #5135: Implement token repair service
For some tokens that were failed to resolve some fields like name, symbol (or it values are empty) we want to retry their resolution.

### PR DESCRIPTION
**Pull request by oa-s**
Closes #5135

### CODE REVIEW

**Comment by hboon** [comment]
@oa-s do you think we can start merging the PRs for the refactor to pipeline first? Otherwise the other PRs are getting too cumbersome to manage?

**Comment by hboon** [comment]
@oa-s tests failing

File: AlphaWallet/Core/Types/Tokens/TokenRepairService.swift [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                case .update(let data, let deletions, let insertions, _):
+                    insertions.map { data[$0] }.forEach { self?.startOperation(for: $0) }
+                    deletions.map { data[$0] }.forEach { self?.cancelOperation(for: $0) }
+                case .error:
+                    self?.cancelAllOperations()
+                }
+            }).store(in: &cancelable)
+    }
+
+    private func buildFetchTokenOrContractOperation(for delegateContract: AddressAndRPCServer) -> AnyPublisher<TokenOrContract, RepairTokenError> {
```
Comment (hboon) [inline AlphaWallet/Core/Types/Tokens/TokenRepairService.swift:77]: Think the name `delegateContract` is wrong.

File: AlphaWallet/Core/Types/Tokens/TokenRepairService.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    private func cancelAllOperations() {
+        for operation in operations {
+            operation.value.cancel()
+        }
+        operations.removeAll()
+    }
+
+    private func cancelOperation(for delegateContract: AddressAndRPCServer) {
```
Comment (hboon) [inline AlphaWallet/Core/Types/Tokens/TokenRepairService.swift:110]: Think the name `delegateContract` is wrong.

File: AlphaWallet/Core/Types/Tokens/TokenRepairService.swift [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            .flatMap { tokenOrContract -> AnyPublisher<TokenOrContract, RepairTokenError> in
+                switch tokenOrContract {
+                case .token, .fungibleTokenComplete: return .just(tokenOrContract)
+                case .ercToken, .delegateContracts, .deletedContracts, .none: return .fail(RepairTokenError.wrongResponse)
+                }
+            }.retry(.randomDelayed(retries: UInt.max, delayBeforeRetry: 10, delayUpperRangeValueFrom0To: 30), scheduler: queue)
+            .eraseToAnyPublisher()
+    }
+
+    private func startOperation(for delegateContract: AddressAndRPCServer) {
```
Comment (hboon) [inline AlphaWallet/Core/Types/Tokens/TokenRepairService.swift:90]: Think the name `delegateContract` is wrong.

File: AlphaWallet/Core/Types/Tokens/TokenRepairService.swift [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    private let queue = DispatchQueue(label: "org.alphawallet.swift.\(String(describing: TokenRepairService.self))", qos: .utility)
+
+    init(tokensDataStore: TokensDataStore, sessionsProvider: SessionsProvider, importToken: ImportToken) {
+        self.tokensDataStore = tokensDataStore
+        self.sessionsProvider = sessionsProvider
+        self.importToken = importToken
+    }
+
+    func start() {
+        let spuriousTokens = tokensDataStore.enabledTokensChangeset(for: Array(sessionsProvider.activeSessions.keys))
```
Comment (hboon) [inline AlphaWallet/Core/Types/Tokens/TokenRepairService.swift:41]: I don't understand how this works?

How does it detect/filter the tokens that need repairing? What is the criteria?

**Comment by oa-s** [comment]
> @oa-s do you think we can start merging the PRs for the refactor to pipeline first? Otherwise the other PRs are getting too cumbersome to manage?

yes,  will start do this. and return to this pr later

**Comment by hboon** [comment]
@oa-s don't forget this one.

**Comment by oa-s** [comment]
continue after https://github.com/AlphaWallet/alpha-wallet-ios/pull/5246

**Comment by hboon** [comment]
@oa-s don't forget this

**Comment by oa-s** [comment]
delaying it for now

**Comment by hboon** [comment]
@oa-s any chance of merging this in but disabled? Lessen the work to keep it updated

**Comment by oa-s** [comment]
i will double check what has left, i remember we had an issue with determining errors when trying to repair a token

**Pull request merged**